stake-pool/single-pool/src/processor.rs

```diff
@@ -0,0 +1,1079 @@
+        let user_lamport_account_info = next_account_info(account_info_iter)?;
+        let clock_info = next_account_info(account_info_iter)?;
+        let clock = &Clock::from_account_info(clock_info)?;
+        let stake_history_info = next_account_info(account_info_iter)?;
+        let token_program_info = next_account_info(account_info_iter)?;
+        let stake_program_info = next_account_info(account_info_iter)?;
+
+        check_pool_stake_address(program_id, vote_account_address, pool_stake_info.key)?;
+        let bump_seed = check_pool_authority_address(
+            program_id,
+            vote_account_address,
+            pool_authority_info.key,
+        )?;
+        check_pool_mint_address(program_id, vote_account_address, pool_mint_info.key)?;
+        check_token_program(token_program_info.key)?;
+        check_stake_program(stake_program_info.key)?;
+
+        // we enforce that our stake is active so depositing/withdrawing during the warmup epoch is impossible
+        let pre_pool_stake = get_active_stake_state(pool_stake_info, clock.epoch)?
+            .delegation
+            .stake;
+        let pre_pool_lamports = pool_stake_info.lamports();
+        msg!("Stake pre merge {}", pre_pool_stake);
+
+        let pre_user_stake = get_active_stake_state(user_stake_info, clock.epoch)?
+            .delegation
+            .stake;
+        let user_unstaked_lamports = user_stake_info
+            .lamports()
+            .checked_sub(pre_user_stake)
+            .ok_or(SinglePoolError::ArithmeticUnderflow)?;
+
+        // merge the user stake account, which is preauthed to us, into the pool stake account
+        // this merge succeeding implicitly validates authority/lockup of the user stake account
+        Self::stake_merge(
+            vote_account_address,
+            user_stake_info.clone(),
+            pool_authority_info.clone(),
+            bump_seed,
+            pool_stake_info.clone(),
+            clock_info.clone(),
+            stake_history_info.clone(),
+            stake_program_info.clone(),
+        )?;
+
+        let post_pool_stake = get_active_stake_state(pool_stake_info, clock.epoch)?
+            .delegation
+            .stake;
+        let post_pool_lamports = pool_stake_info.lamports();
+        msg!("Stake post merge {}", post_pool_stake);
+
+        // all lamports added, as a lamport difference, both stake and excess rent
+        let lamports_added = post_pool_lamports
+            .checked_sub(pre_pool_lamports)
+            .ok_or(SinglePoolError::ArithmeticUnderflow)?;
+
+        // stake lamports added, as a stake difference
+        let stake_added = post_pool_stake
+            .checked_sub(pre_pool_stake)
+            .ok_or(SinglePoolError::ArithmeticUnderflow)?;
+
+        // excess rent, as all lamports less stake
+        let leftover_rent = lamports_added
+            .checked_sub(stake_added)
+            .ok_or(SinglePoolError::ArithmeticUnderflow)?;
```

Comment:
This is the safest behavior, but I want to walk through an alternative.

What if we have `leftover_rent = post_pool_lamports - post_pool_stake - stake_meta.stake_rent - minimum_delegation`? With that logic, if someone just transfers lamports into the stake pool expecting pool tokens, the following depositor can reclaim those lamports.

Most importantly, this means that the stake account never carries around extra lamports, and if someone messes up, they can always reclaim their funds by depositing immediately after. Or, more likely, someone else will. Otherwise, it's possible to just have unused lamports bricked in these accounts. What do you think?